### DISCUSSION TITLE
Consistent OS window, tab, and window indices?

### DISCUSSION THREAD

**Discussion by okuuva** [kovidgoyal/kitty#5583] [Q&A]
I'm in the process of replacing my workflow using Alacritty + iTerm2 + tmux combination with just kitty and nothing else. So far I've managed get it almost ready with Hammerspoon powered dropdown terminal and session files. While session files are good enough replacement for server/client style of session management a'la iTerm and tmux I'd still like to make it "dynamic" so that every OS window, tab and window would resume where I left off if for whatever reason I need to restart kitty tweaking some settings being the most frequent reason.

I had an idea how to do that using the combination of `shell` setting in kitty.conf and abduco. What I want to do is following:

`shell abduco -fA kitty-<os_win_index>-<tab_index>-<window_index> fish --login --interactive`

Where indices would be consistent between "sessions" (i.e. kitty restarts). They don't have to be indices to be precise, as long as they would stay the same every time kitty is completely restarted. That way the second window of the third tab in the fourth OS window would always get the same abduco session name (in this case something like `kitty-4-3-2`) and would automatically continue where it left off before restart as long as OS windows, tabs and windows are opened in the same order they were before restarting.

Granted, this would be somewhat flimsy setup because like said the OS windows, tabs and windows would need to be launched in the exact same order as in previous "session" but it would work in my use case.

So the question is: is there any such indices / ids for OS windows, tabs and windows? And if there is is there any way to use them in kitty.conf or would I have to use some kitten for that?

**Comment by page-down** [comment]
> ... I need to restart kitty tweaking some settings ... most frequent reason.

Most of the kitty configurations can be applied by reloading the `kitty.conf`(ctrl+shift+f5).
https://sw.kovidgoyal.net/kitty/conf/#shortcut-kitty.Reload-kitty.conf

> ... I've managed get it almost ready with ... session files.
> ... I'd still like to make it "dynamic" ...

Although you can learn about kitty windows via the environment variable `KITTY_WINDOW_ID` and the remote control `ls` command (with wrapper script), this sounds rather unreliable.

If you need a simple solution, the appropriate steps I can think of are:

- Do not configure `shell` with some terminal multiplexer, because some features in kitty expect to run a real shell non-interactively.
- Bind some shortcuts (e.g. new window, new tab) to run with a wrapper script (get the session name via env vars or args).
- Generate a unique ID for the window in the wrapper script.
- Write a kitten that exports the current window tree with the associated unique ID before exiting. A temporary session file can also be generated in the meantime.
- Run `kitty --session` to load the session file and delete it after start.

I tried to use `abduco` with a few lines of wrapper script and a few dozen lines of kitten to achieve the `dynamic recovery` you describe, but `abduco` has more problems than is acceptable for daily use.

Related GitHub issue (kitty daemon mode):
https://github.com/kovidgoyal/kitty/issues/391

**Reply by okuuva** [reply]
>Most of the kitty configurations can be applied by reloading the kitty.conf(ctrl+shift+f5).
https://sw.kovidgoyal.net/kitty/conf/#shortcut-kitty.Reload-kitty.conf

I know, but some of the settings can't be hotloaded. And I do like how terminal session continues from where I left off after the terminal restart anyway, even if the session isn't restored after restarting the computer.

Thanks for the quick response!
> * Do not configure shell with some terminal multiplexer...

Not planning to. `abduco` in itself isn't a terminal multiplexer, just a thin wrapper to allow detaching from and reattaching into running programs. By default it launches `dvtm` terminal multiplexer if no program is given as a startup argument but I'm currently using it to launch a login terminal and leaving window and tab management to kitty.

>* Bind some shortcuts (e.g. new window, new tab) to run with a wrapper script (get the session name via env vars or args).
>* Generate a unique ID for the window in the wrapper script.
>* Write a kitten that exports the current window tree with the associated unique ID before exiting. A temporary session file can also be generated in the meantime.
>* Run kitty --session to load the session file and delete it after start.

This is pretty much what I had in mind, thanks! Would've been easier if the IDs were accessible straight from `kitty.conf` but I knew it was a long shot anyway.

>* Write a kitten that exports the current window tree with the associated unique ID before exiting. A temporary session file can also be generated in the meantime.

Could you point me to right direction on how to run a specific kitten on exit and how to generate a temporary session file? I know there must be instructions and examples somewhere but so far I've been very bad at finding those from old issues and the homepage 😓

>Related GitHub issue (kitty daemon mode):
https://github.com/kovidgoyal/kitty/issues/391

Now that one I did find and read through it. This whole Q&A is just a workaround in an effort to get my workflow as close to as it was before with tmux before we (hopefully) get `kittyd` at some point. I could take a look at it myself but since I'm really not fluent in C I doubt I'm gonna be much of a help in it.

**Reply by kovidgoyal** [reply]
If you want to run code on exit from a window, use the watchers function. In it check if it is the last window that is closing to detect application exit. And there already exist scripts to generate session files from current kitty state using remote control, a bit of googling should find them. But note that a session file can restore window/tab layout and restart programs but it wont restore actual state of those programs or the scrollback buffer, etc.
